README.md
# Google Cloud BigQuery &middot;  [![NPM](https://img.shields.io/npm/v/google-cloud-bigquery.svg?style=flat)](https://www.npmjs.com/package/google-cloud-bigquery) [![Tests](https://travis-ci.org/nicolasdao/google-cloud-bigquery.svg?branch=master)](https://travis-ci.org/nicolasdao/google-cloud-bigquery) [![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause) [![Neap](https://neap.co/img/made_by_neap.svg)](#this-is-what-we-re-up-to)
__*Google Cloud BigQuery*__ is a node.js package to maintain BigQuery table, either explicitely or using a Google Cloud Storage (including automatically updating the tables' schema).

# Table of Contents

> * [Install](#install) 
> * [Getting started](#getting-started) 
>	- [Prerequisite](#prerequisite)
>	- [Four ways to create a client](#four-ways-to-create-a-client)
>		- [User the hosting identity](#user-the-hosting-identity)
>		- [Using a `service-account.json`](#using-a-service-accountjson)
>		- [Using explicit credentials](#using-explicit-credentials)
>		- [Using environment variables](#using-environment-variables)
>	- [Creating a new table](#creating-a-new-table)
>	- [Inserting data](#inserting-data)
>	- [Getting data](#getting-data)
>	- [Updating the table's schema](#updating-the-tables-schema)
> * [Best Practices - Reliability & Performances](#best-practices---reliability--performances)
> * [Useful Code Snippets](#snippets-to-put-it-all-together)
> * [About Neap](#this-is-what-we-re-up-to)
> * [License](#license)


# Install
```
npm i google-cloud-bigquery
```

# Getting started
## Prerequisite

Before using this package, you must first:

1. Have a Google Cloud Account.
2. Have a both a BigQuery DB and a Bucket in the same region (the bucket is only in case you wish to maintain BigQuery schema using data stored a Google Cloud Storage). As of December 2018, BigQuery is only supported in the following locations:
	- asia-northeast1 (Tokyo)
	- asia-east1 (Taiwan)
	- asia-southeast1 (Singapore)
	- australia-southeast1 (Sydney)
	- europe-north1 (Finland)
	- europe-west2 (London)
	- us-east4 (Northern Virginia)
	- eu (Multi regions in the EU)
	- us (Multi regions in the US)
3. Have a Service Account set up with the following 2 roles:
	- `roles/bigquery.admin`
	- `roles/storage.objectAdmin` (only in case you wish to maintain BigQuery schema using data stored a Google Cloud Storage)
4. Get the JSON keys file for that Service Account above.
5. Save that JSON key into a `service-account.json` file (make sure it is located under a path that is accessible to your app), or save the following properties to either manually set up the client or set up environment variables:
	- `project_id`
	- `client_email`
	- `private_key`
6. Modify the `service-account.json` above by adding a new `location_id` property with the location ID of your BigQuery service (e.g., `australia-southeast1`).

## Four ways to create a client

This library supports four different ways to create a client. The first method is the recommended way:
1. [User the hosting identity](#user-the-hosting-identity)
2. [Using a `service-account.json`](#using-a-service-accountjson)
3. [Using explicit credentials](#using-explicit-credentials)
4. [Using environment variables](#using-environment-variables)

### User the hosting identity

```js
const { client } = require('google-cloud-bigquery')

const bigQuery = client.new()
```

In this case, the package fetches the credentials automatically. It will try three different techniques to get those data, and if none of them work, an error is thrown. Those techniques are:
1. If the code is hosted on GCP (e.g., Cloud Compute, App Engine, Cloud Function or Cloud Run) then the credentials are extracted from the service account associated with the GCP service.
2. If the `GOOGLE_APPLICATION_CREDENTIALS` environment variable exists, its value is supposed to be the path to a service account JSON key file on the hosting machine.
3. If the `~/.config/gcloud/application_default_credentials.json` file exists, then the credentials it contains are used (more about setting that file up below).

When developing on your local environment, use either #2 or #3. #3 is equivalent to being invited by the SysAdmin to the project and granted specific privileges. To set up `~/.config/gcloud/application_default_credentials.json`, follow those steps:

- Make sure you have a Google account that can access both the GCP project and the resources you need on GCP. 
- Install the `GCloud CLI` on your environment.
- Execute the following commands:
	```
	gcloud auth login
	gcloud config set project <YOUR_GCP_PROJECT_HERE>
	gcloud auth application-default login
	```
	The first command logs you in. The second command sets the `<YOUR_GCP_PROJECT_HERE>` as your default project. Finally, the third command creates a new `~/.config/gcloud/application_default_credentials.json` file with the credentials you need for the `<YOUR_GCP_PROJECT_HERE>` project.

### Using a `service-account.json`

We assume that you have created a Service Account in your Google Cloud Account (using IAM) and that you've downloaded a `service-account.json` (the name of the file does not matter as long as it is a valid json file). The first way to create a client is to provide the path to that `service-account.json` as shown in the following example:

```js
const { join } = require('path')
const { client } = require('google-cloud-bigquery')

const bigQuery = client.new({ jsonKeyFile: join(__dirname, './service-account.json') })
```


### Using explicit credentials

This method is similar to the previous one. You should have dowloaded a `service-account.json`, but instead of providing its path, you provide some of its details explicitly:

```js
const { client } = require('google-cloud-bigquery')

const bigQuery = client.new({
	credentials: {
		project_id: 'test', 
		location_id: 'australia-southeast1',
		client_email:'something-1234@your-project-id.iam.gserviceaccount.com', 
		private_key: '-----BEGIN PRIVATE KEY-----\n123456789-----END PRIVATE KEY-----\n'
	}
})
```

> All those details should be coming from the `service-account.json` you downloaded in the [Prerequisite](#prerequisite) step.

### Using environment variables

```js
const { client } = require('google-cloud-bigquery')

const bigQuery = client.new()
```

The above will only work if all the following environment variables are set:
- `GOOGLE_CLOUD_BIGQUERY_PROJECT_ID` or `GOOGLE_CLOUD_PROJECT_ID`
- `GOOGLE_CLOUD_BIGQUERY_LOCATION_ID` or `GOOGLE_CLOUD_LOCATION_ID`
- `GOOGLE_CLOUD_BIGQUERY_CLIENT_EMAIL` or `GOOGLE_CLOUD_CLIENT_EMAIL`
- `GOOGLE_CLOUD_BIGQUERY_PRIVATE_KEY` or `GOOGLE_CLOUD_PRIVATE_KEY`

> WARNING: If you're using [NPM's `dotenv`](https://www.npmjs.com/package/dotenv), wrap your PRIVATE_KEY between double-quotes, otherwise some characters are escaped which corrupts the key.

## Creating a new table

```js
const { join } = require('path')
const { client } = require('google-cloud-bigquery')

const bigQuery = client.new({ jsonKeyFile: join(__dirname, './service-account.json') })

// Assumes that 'your-dataset-id' already exists
const db = bigQuery.db.get('your-dataset-id')
const userTbl = db.table('user')

userTbl.exists()
	.then(yes => yes 
		? console.log(`Table '${userTbl.name}' already exists in DB '${db.name}'`)
		: userTbl.create.new({ 
			schema: {
				id: 'integer',
				username: 'string',
				friends: [{
					id: 'integer',
					username: 'string',
					score: 'float'
				}],
				country: {
					code: 'string',
					name: 'string'
				},
				married: 'boolean',
				tags:['string'],
				inserted_date: 'timestamp'
			} 
		}).then(() => console.log(`Table '${userTbl.name}' successfully added to DB '${db.name}'`)))
```

## Inserting data

```js
userTbl.insert.values({ data:[{
		id: 1,
		username: 'Nicolas',
		inserted_date: new Date()
	}, {
		id: 2,
		username: 'Brendan',
		country: {
			code: 'AU',
			name: 'Australia'
		},
		friends:[{
			id: 1,
			username: 'Nicolas',
			score: 0.87
		}, {
			id: 3,
			username: 'Boris',
			score: 0.9
		}],
		inserted_date: new Date()
	}, {
		id: '3',
		username: 'Boris',
		tags:['admin',1],
		inserted_date: Date.now()/1000
	}]
})
```

### IMPORTANT NOTE ABOUT QUOTAS AND LIMITS

Notice that the `data` input accept both single objects or array of objects. Though BigQuery can ingest up to 10,000 rows per request and 100,000 rows per seconds, it is recommended to keep the maximum amount of rows per request to 500. You can read more about the quotas and limits at [https://cloud.google.com/bigquery/quotas#streaming_inserts](https://cloud.google.com/bigquery/quotas#streaming_inserts).

To prevent inserting more than 500 rows per request, you can either code it yourself, or rely on our own implementation using the `safeMode` flag as follow:

```js
userTbl.insert.values({ data: lotsOfUsers, safeMode: true })
	.then(() => console.log(`All users inserted`))
```

This `safeMode` flag will check that there is less than 500 items in the _lotsOfUsers_ array. If there are more than 500 items, the array is broken down in batches of 500 items which are then inserted sequentially. That means that if you're inserting 5000 users, there will be 10 sequential request of 500 users.


## Getting data

```js
db.query.execute({ 
	sql:`select * from ${db.name}.${userTbl.name} where id = @id`, 
	params: { id: 2 } 
})

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.0.0](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.5.2...v1.0.0) (2020-08-20)


### Features

* Add support for using the application default credentials ([cd3693c](https://github.com/nicolasdao/google-cloud-bigquery/commit/cd3693c3388a5638752f2094a1adc78724aa01a2))

### [0.5.2](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.5.1...v0.5.2) (2020-08-01)


### Bug Fixes

* Creating a client without any argument fails ([c1e87f7](https://github.com/nicolasdao/google-cloud-bigquery/commit/c1e87f76406702adc83f6499a752e44ffd537232))

### [0.5.1](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.5.0...v0.5.1) (2020-08-01)


### Bug Fixes

* Invalid logic when deciding which credential strategy to choose to create a client ([bc73bce](https://github.com/nicolasdao/google-cloud-bigquery/commit/bc73bce570622565fb32070a4c94d2750a6a5b27))
* Vulnerabilities ([7f9640c](https://github.com/nicolasdao/google-cloud-bigquery/commit/7f9640cc0258a064b895fa0e834ae8a5fd480d1c))

## [0.5.0](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.4.3...v0.5.0) (2020-08-01)


### Features

* Add support for creating a client using explicit creds or environment variables ([f0da246](https://github.com/nicolasdao/google-cloud-bigquery/commit/f0da246edda1290aea375fc8eeaf28707116872d))

### [0.4.3](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.4.2...v0.4.3) (2020-07-16)


### Bug Fixes

* Vulnerabilities ([46119a5](https://github.com/nicolasdao/google-cloud-bigquery/commit/46119a51c386ea1b3e1fbc00f37320d8b92c1dae))

<a name="0.4.2"></a>
## [0.4.2](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.4.1...v0.4.2) (2019-02-16)



<a name="0.4.1"></a>
## [0.4.1](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.4.0...v0.4.1) (2019-02-16)


### Bug Fixes

* Missing error message for bad queries + incorrect type casting when number string are passed to queries ([0b4541d](https://github.com/nicolasdao/google-cloud-bigquery/commit/0b4541d))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.3.2...v0.4.0) (2019-02-07)


### Features

* Add support for reliable safe mode for insert of big number of rows at once ([6707029](https://github.com/nicolasdao/google-cloud-bigquery/commit/6707029))



<a name="0.3.2"></a>
## [0.3.2](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.3.1...v0.3.2) (2019-02-07)


### Bug Fixes

* Inserts that exceed the streaming insert quotas limit do not throw any errors ([9f9d664](https://github.com/nicolasdao/google-cloud-bigquery/commit/9f9d664))



<a name="0.3.1"></a>
## [0.3.1](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.3.0...v0.3.1) (2018-12-10)


### Features

* Throw errors if bad http status ([a838c3c](https://github.com/nicolasdao/google-cloud-bigquery/commit/a838c3c))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.2.11...v0.3.0) (2018-12-10)


### Features

* Improve the robusness with retry mechanism ([d8c9a43](https://github.com/nicolasdao/google-cloud-bigquery/commit/d8c9a43))



<a name="0.2.11"></a>
## [0.2.11](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.2.10...v0.2.11) (2018-12-10)



<a name="0.2.10"></a>
## [0.2.10](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.2.9...v0.2.10) (2018-12-10)


### Features

* Add support for safeMode when inserting huge amount of rows ([579c959](https://github.com/nicolasdao/google-cloud-bigquery/commit/579c959))



<a name="0.2.9"></a>
## [0.2.9](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.2.8...v0.2.9) (2018-12-09)



<a name="0.2.8"></a>
## [0.2.8](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.2.7...v0.2.8) (2018-12-09)



<a name="0.2.7"></a>
## [0.2.7](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.2.6...v0.2.7) (2018-12-04)


### Features

* Add support for getting the table and db name straight from the object themselves ([098134c](https://github.com/nicolasdao/google-cloud-bigquery/commit/098134c))



<a name="0.2.6"></a>
## [0.2.6](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.2.5...v0.2.6) (2018-12-03)



<a name="0.2.5"></a>
## [0.2.5](https://github.com/nicolasdao/google-cloud-bigquery/compare/v0.2.4...v0.2.5) (2018-12-03)


### Features


src/format.js
/**
 * Copyright (c) 2018, Neap Pty Ltd.
 * All rights reserved.
 * 
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
*/

const NULL_TIMESTAMP = '1970-01-01T00:00:00.000Z'
const NULL_FLOAT = 5e-16
const NULL_INT = -1
const NULL_STRING = 'NULL'

/**
 * Remove all properties of an object that represent numbers. This is required before addind data to BigQuery
 * as BigQuery does not support number fields.
 * 
 * @param  {Object} obj e.g., { "123": { hello: "world" } }
 * @return {Object}     e.g., { "#123": { hello: "world" } }
 */
const cleanData = (obj) => {
	if (!obj || obj instanceof Date || typeof(obj) != 'object')
		return obj 

	if (Array.isArray(obj))
		return obj.map(cleanData)

	const keys = Object.keys(obj)
	if (keys.length == 0)
		return obj 

	return keys.reduce((acc,key) => {
		const v = obj[key]
		const newVal = cleanData(v)
		const newKey = /^[0-9]*\.{0,1}[0-9]+$/.test(key) ? `#${key}` : key
		acc[newKey] = newVal
		return acc
	}, {})
}

const _stringify = (v,options={}) => options.mode == 'exact' ? `string:${v}` : `${v}`

const _convertValToTimestamp = (fieldType, fieldIsDate, fieldValue, options={}) => {
	const { mode } = options
	const exactMode = mode == 'exact'
	if (fieldType == 'string' || fieldType == 'number') {
		const d = new Date(fieldValue)
		if (d.toString().toLowerCase() == 'invalid date')
			return exactMode ? NULL_TIMESTAMP : null
		else
			return d.toISOString()
	} else if (fieldIsDate) 
		return fieldValue.toISOString()
	else
		return exactMode ? NULL_TIMESTAMP : null
}

const _convertFieldToFloat = (fieldType, fieldIsArray, fieldIsDate, fieldIsObject, fieldValue, options={}) => {
	const { mode } = options
	const exactMode = mode == 'exact'

	if (fieldIsDate)
		return exactMode ? fieldValue.getTime() + NULL_FLOAT : fieldValue.getTime()
	else if (fieldIsArray || fieldIsObject)
		return exactMode ? NULL_FLOAT : null
	else if (fieldType == 'string') {
		const isNumber = /^[0-9]*\.{0,1}[0-9]+$/.test(fieldValue)
		if (isNumber) {
			const n = fieldValue * 1
			if (n - Math.round(n))
				return n
			else
				return exactMode ? n + NULL_FLOAT : null
		}
	} else if (fieldType == 'number') {
		if ((fieldValue - Math.round(fieldValue)) || !exactMode)
			return fieldValue
		else
			return fieldValue + NULL_FLOAT
	} else if (fieldType == 'boolean') {
		if (exactMode) 
			return fieldValue ? 1+NULL_FLOAT : NULL_FLOAT
		else 
			return fieldValue ? 1 : 0
	}else
		return exactMode ? NULL_FLOAT : null
}

const _allowedSchemaTypes = { 'string': true, 'integer': true, 'boolean': true, 'object': true, 'timestamp': true, 'float': true }
const fitToSchema = (obj={}, schema={}, options={}) => {
	const { mode } = options
	const exactMode = mode == 'exact'
	const nullString = exactMode ? NULL_STRING : ''
	const keys = Object.keys(schema)
	if (keys.length == 0)
		return {}
	
	const newObj = keys.reduce((acc,key) => {
		const v = obj[key]
		if (!exactMode && v === undefined)
			return acc
		const schemaType = schema[key]
		let sType
		const schemaTypeIsArray = Array.isArray(schemaType)
		const schemaTypeIsObject = typeof(schemaType) == 'object'
		if (schemaTypeIsArray) {
			if (!schemaType[0])
				throw new Error(`Invalid type. Field '${key}' is an array of empty type (ref. #01).`)

			const t = typeof(schemaType[0]) == 'object' ? 'object' : schemaType[0]
			sType = `${t}`.toLowerCase()
			if (!_allowedSchemaTypes[sType])
				throw new Error(`Field '${key}' uses an unsupported type ${sType}. Only allowed types are: ${Object.keys(_allowedSchemaTypes)} (ref. #02)`)

			sType = `array-${sType}`
		} else {
			sType = schemaTypeIsObject ? 'object' : `${schemaType}`.toLowerCase()
			if (!_allowedSchemaTypes[sType])
				throw new Error(`Field '${key}' uses an unsupported type ${sType}. Only allowed types are: ${Object.keys(_allowedSchemaTypes)} (ref. #03)`)
		}

		const fieldType = typeof(v)
		const fieldIsArray = Array.isArray(v)
		const fieldIsDate = v instanceof Date
		const fieldIsObject = fieldType == 'object'

		if (!v) {
			if (!exactMode) {
				if (v === 0) {
					if (sType == 'string')
						acc[key] = '0'
					else if (sType == 'boolean')
						acc[key] = false
					else if (sType == 'integer' || sType == 'float')
						acc[key] = 0
					else 
						acc[key] = null
				} else if (v === '') {
					if (sType == 'string')
						acc[key] = ''
					else if (sType == 'boolean')
						acc[key] = false
					else 
						acc[key] = null
				} else if (v === false) {
					if (sType == 'boolean')
						acc[key] = false
					else 
						acc[key] = null
				} else
					acc[key] = null

				return acc
			}
			if (sType == 'string') 
				acc[key] = _stringify(nullString,options)
			else if (sType == 'integer') 
				acc[key] = v === 0 ? 0 : NULL_INT
			else if (sType == 'float') 
				acc[key] = NULL_FLOAT
			else if (sType == 'boolean') 
				acc[key] = false
			else if (sType == 'timestamp') 
				acc[key] = NULL_TIMESTAMP
			else if (schemaTypeIsArray) {
				const schemaArrayType = schemaType[0]
				if (sType == 'array-string')
					acc[key] = [_stringify(nullString,options)]
				else if (sType == 'array-integer')
					acc[key] = [NULL_INT]
				else if (sType == 'array-boolean')
					acc[key] = [false]
				else if (sType == 'array-timestamp')
					acc[key] = [NULL_TIMESTAMP]
				else if (sType == 'array-float')
					acc[key] = [NULL_FLOAT]
				else if (sType == 'array-object')
					acc[key] = [fitToSchema({}, schemaArrayType, { mode, isNull: true })]
				else
					throw new Error(`Field '${key}' uses an unsupported type ${sType} (ref. #04).`)
			} else if (sType == 'object') 
				acc[key] = fitToSchema({}, schemaType, { mode, isNull: true })
			else
				throw new Error(`Field '${key}' uses an unsupported type ${sType} (ref. #05).`)
		}
		else if (v){
			if (sType == 'string') {
				const newVal = 
					fieldIsDate ? _stringify(v.toISOString(),options) : 
						fieldIsObject ? 'Object' : _stringify(v,options)
				acc[key] = newVal
			} else if (sType == 'integer') {
				const newVal = 
					fieldType == 'string' && /^[0-9]*\.{0,1}[0-9]+$/.test(v) ? Math.round(v*1) : 
						fieldType == 'boolean' ? (v ? 1 : 0) :
							fieldIsDate ? v.getTime() : 
								fieldType == 'number' ? Math.round(v) : 
									exactMode ?	NULL_INT : null

				acc[key] = newVal
			} else if (sType == 'float') {
				acc[key] = _convertFieldToFloat(fieldType, fieldIsArray, fieldIsDate, fieldIsObject, v, options)
			}else if (sType == 'boolean') {
				const newVal = 
					fieldType == 'string' ? (v ? true : false) : 
						fieldType == 'number' ? (v ? true : false) : 
							fieldIsArray || fieldIsObject || fieldIsDate ? true : false

				acc[key] = newVal
			} else if (sType == 'timestamp') {
				acc[key] = _convertValToTimestamp(fieldType, fieldIsDate, v, options)
			} else if (schemaTypeIsArray) {
				const schemaArrayType = schemaType[0]
				const newVal = 
					fieldType == 'string' && sType == 'array-string' ? [_stringify(v,options)] : 
						fieldType == 'string' && sType == 'array-integer' ? (/^[0-9]*\.{0,1}[0-9]+$/.test(v) ? [Math.round(v*1)] : [NULL_INT]) :  
							fieldType == 'string' && sType == 'array-float' ? [_convertFieldToFloat(fieldType, fieldIsArray, fieldIsDate, fieldIsObject, v, options)] :  
								fieldType == 'string' && sType == 'array-boolean' ? [true] : 
									fieldType == 'string' && sType == 'array-object' ? [fitToSchema({}, schemaArrayType, { mode, isNull: true })] : 
										fieldType == 'string' && sType == 'array-timestamp' ? [_convertValToTimestamp(fieldType, fieldIsDate, v, options)] : 
											fieldType == 'number' && sType == 'array-string' ? [_stringify(v,options)] : 
												fieldType == 'number' && sType == 'array-integer' ? [Math.round(v)] :  
													fieldType == 'number' && sType == 'array-float' ? [_convertFieldToFloat(fieldType, fieldIsArray, fieldIsDate, fieldIsObject, v, options)] :  
														fieldType == 'number' && sType == 'array-boolean' ? (v ? [true] : [false]) : 
															fieldType == 'number' && sType == 'array-object' ? [fitToSchema({}, schemaArrayType, { mode, isNull: true })] : 
																fieldType == 'number' && sType == 'array-timestamp' ? [_convertValToTimestamp(fieldType, fieldIsDate, v, options)] : 
																	fieldType == 'boolean' && sType == 'array-string' ? (v ? [_stringify('true',options)] : [_stringify('false',options)]) : 
																		fieldType == 'boolean' && sType == 'array-integer' ? (v ? [1] : [0]) :  
																			fieldType == 'boolean' && sType == 'array-float' ? [_convertFieldToFloat(fieldType, fieldIsArray, fieldIsDate, fieldIsObject, v, options)] :  
																				fieldType == 'boolean' && sType == 'array-boolean' ? (v ? [true] : [false]) : 
																					fieldType == 'boolean' && sType == 'array-object' ? [fitToSchema({}, schemaArrayType, { mode, isNull: true })] : 
																						fieldType == 'boolean' && sType == 'array-timestamp' ? [_convertValToTimestamp(fieldType, fieldIsDate, v, options)] : 
																							fieldIsDate && sType == 'array-string' ? [_stringify(v.toISOString(),options)] : 
																								fieldIsDate && sType == 'array-integer' ? [v.getTime()] :  
																									fieldIsDate && sType == 'array-float' ? [_convertFieldToFloat(fieldType, fieldIsArray, fieldIsDate, fieldIsObject, v, options)] :  
																										fieldIsDate && sType == 'array-boolean' ? [true] : 
																											fieldIsDate && sType == 'array-object' ? [fitToSchema({}, schemaArrayType, { mode, isNull: true })] : 
																												fieldIsDate && sType == 'array-timestamp' ? [_convertValToTimestamp(fieldType, fieldIsDate, v, options)] : 
																													fieldIsArray ? _getNewArrayVal(v,sType,schemaArrayType,key,options) : 
																														fieldType == 'object' && sType == 'array-string' ? [_stringify(nullString,options)] : 
																															fieldType == 'object' && sType == 'array-integer' ? [NULL_INT] :  
																																fieldType == 'object' && sType == 'array-float' ? [_convertFieldToFloat(fieldType, fieldIsArray, fieldIsDate, fieldIsObject, v, options)] :  
																																	fieldType == 'object' && sType == 'array-boolean' ? [false] : 
																																		fieldType == 'object' && sType == 'array-object' ? [fitToSchema(v, schemaArrayType, { mode, isNull: false })] : 
																																			fieldType == 'object' && sType == 'array-timestamp' ? [_convertValToTimestamp(fieldType, fieldIsDate, v, options)] : null 

				if (!newVal)
					throw new Error(`Field '${key}' uses an unsupported type ${sType} (ref. #06).`)

				acc[key] = newVal
			} else if (sType == 'object') {
				const newVal = !fieldIsArray && fieldIsObject ? v : {}
				const opts = Object.keys(newVal).some(x => x) ? { mode, isNull: false } : { mode, isNull: true }
				acc[key] = fitToSchema(newVal, schemaType, opts)
			}
			else
				throw new Error(`Field '${key}' uses an unsupported type ${sType} (ref. #07).`)
		}

		return acc
	}, options.isNull === undefined || !exactMode ? {} : { _isNull: options.isNull ? true : false })
	
	if (Object.keys(newObj).some(x => x))
		return newObj
	else
		return null
}

const _getNewArrayVal = (arrayVal=[], sType, schemaType, key, options={}) => {
	const { mode } = options
	const nullString = mode == 'exact' ? NULL_STRING : ''
	const a = arrayVal.some(x => x) ? arrayVal : [_stringify(nullString,options)]
	const result = a.map(x => {
		const v = 
			!x && typeof(x) == 'string' ? nullString : 
				!x && typeof(x) == 'number' ? 0 : 
					!x ? nullString : x

		const fieldType = typeof(v)
		const fieldIsDate = v instanceof Date

		const newVal = 
			fieldType == 'string' && sType == 'array-string' ? _stringify(v,options) : 
				fieldType == 'string' && sType == 'array-integer' ? (/^[0-9]*\.{0,1}[0-9]+$/.test(v) ? (Math.round(v*1)) : NULL_INT) :  
					fieldType == 'string' && sType == 'array-float' ? _convertFieldToFloat(fieldType, false, fieldIsDate, false, v, options) :  
						fieldType == 'string' && sType == 'array-boolean' ? true : 
							fieldType == 'string' && sType == 'array-object' ? fitToSchema({}, schemaType, { mode, isNull: true }) : 
								fieldType == 'string' && sType == 'array-timestamp' ? _convertValToTimestamp(fieldType, fieldIsDate, v, options) : 
									fieldType == 'number' && sType == 'array-string' ? _stringify(v,options) : 
										fieldType == 'number' && sType == 'array-integer' ? Math.round(v) :  
											fieldType == 'number' && sType == 'array-float' ? _convertFieldToFloat(fieldType, false, fieldIsDate, false, v, options) :  
												fieldType == 'number' && sType == 'array-boolean' ? (v ? true : false) : 
													fieldType == 'number' && sType == 'array-object' ? fitToSchema({}, schemaType, { mode, isNull: true }) : 
														fieldType == 'number' && sType == 'array-timestamp' ? _convertValToTimestamp(fieldType, fieldIsDate, v, options) : 
															fieldType == 'boolean' && sType == 'array-string' ? (v ? _stringify('true',options) : _stringify('false',options)) : 
																fieldType == 'boolean' && sType == 'array-integer' ? (v ? 1 : 0) :  
																	fieldType == 'boolean' && sType == 'array-float' ? _convertFieldToFloat(fieldType, false, fieldIsDate, false, v, options) :  
																		fieldType == 'boolean' && sType == 'array-boolean' ? (v ? true : false) : 
																			fieldType == 'boolean' && sType == 'array-object' ? fitToSchema({}, schemaType, { mode, isNull: true }) : 
																				fieldType == 'boolean' && sType == 'array-timestamp' ? _convertValToTimestamp(fieldType, fieldIsDate, v, options) : 
																					fieldIsDate && sType == 'array-string' ? _stringify(v.toISOString(),options) : 
																						fieldIsDate && sType == 'array-integer' ? v.getTime() :  
																							fieldIsDate && sType == 'array-float' ? _convertFieldToFloat(fieldType, false, fieldIsDate, false, v, options) :  
																								fieldIsDate && sType == 'array-boolean' ? true : 
																									fieldIsDate && sType == 'array-object' ? fitToSchema({}, schemaType, { mode, isNull: true }) : 
																										fieldIsDate && sType == 'array-timestamp' ? _convertValToTimestamp(fieldType, fieldIsDate, v, options) : 
																											fieldType == 'object' && sType == 'array-string' ? _stringify(nullString,options) : 
																												fieldType == 'object' && sType == 'array-integer' ? NULL_INT :  
																													fieldType == 'object' && sType == 'array-float' ? _convertFieldToFloat(fieldType, false, fieldIsDate, false, v, options) : 
																														fieldType == 'object' && sType == 'array-boolean' ? false : 
																															fieldType == 'object' && sType == 'array-object' ? fitToSchema(v, schemaType, { mode, isNull: false }) : 
																																fieldType == 'object' && sType == 'array-timestamp' ? _convertValToTimestamp(fieldType, fieldIsDate, v, options) : null 

		if (mode == 'exact' && !newVal) 
			throw new Error(`Field '${key}' uses an unsupported type ${sType} (ref. #08).`)

		return newVal
	})

	if (sType == 'array-object')
		return result.filter(x => x)
	else
		return result
}

const _primitiveTypes = { 'number': 'INTEGER', 'integer': 'INTEGER', 'float': 'FLOAT', 'timestamp': 'TIMESTAMP', 'boolean': 'BOOLEAN', 'string': 'STRING' }
/**
 * [description]
 * @param  {Object}  schema [description]
 * @return {[Field]} results.fields       
 * @return {String}  results.fields.name      Field's name
 * @return {String}  results.fields.type      One of the following values: 'INTEGER', 'FLOAT', 'TIMESTAMP', 'BOOLEAN', 'STRING', 'RECORD'
 * @return {String}  results.fields.mode      One of the following values: 'NULLABLE', 'REPEATED', 'REQUIRED'
 * @return {[Field]} results.fields.fields    If 'type' is 'RECORD', then these are the fields of the record 
 */
const schemaToFields = (schema={}) => {
	const keys = Object.keys(schema)
	if (!keys.some(x => x))
		return { fields:[] }

	return keys.reduce((acc,name) => {
		const v = schema[name]
		const mode = Array.isArray(v) ? 'REPEATED' : 'NULLABLE'
		let type = _primitiveTypes[`${v}`.toLowerCase()]
		let fields
		if (!type) {
			if (mode == 'NULLABLE' && typeof(v) != 'object')
				throw new Error(`Invalid schema type. Field '${name}' is using the unsupported type '${v}'`)

			if (mode == 'NULLABLE') {
				type = 'RECORD'
				fields = schemaToFields(v).fields
			} else {
				const arg = v[0]
				type = _primitiveTypes[`${v}`.toLowerCase()]
				if (!type) {
					if (typeof(arg) != 'object')
						throw new Error(`Invalid schema type. Field '${name}' is using the unsupported type '[${v}]'`)
					type = 'RECORD'
					fields = schemaToFields(arg).fields
				}
			}
		}
		let field = { name, type, mode }
		if (fields)
			field.fields = fields
		acc.fields.push(field)
		return acc
	}, { fields:[] })
}

const fieldsToSchema = fields => {
	fields = fields || []
	return fields.reduce((acc, { name, type, mode, fields }) => {
		const t = type.toLowerCase().trim()
		const m = mode.toLowerCase().trim()
		if (t != 'record' && m != 'repeated')
			acc[name] = t 
		else if (t == 'record') {
			const v = fieldsToSchema(fields)
			if (m == 'repeated')
				acc[name] = [v]
			else
				acc[name] = v	
		} else if (m == 'repeated') 
			acc[name] = [t]

		return acc
	}, {})
}

const bigQueryResultToJson = (data={}) => {
	if (data && data.schema && data.schema.fields && data.rows) {
		const schema = data.schema.fields.map(({ name, type, fields }) => {
			if (type == 'INTEGER' || type == 'DECIMAL' || type == 'FLOAT')
				return { name, convert: x => x * 1 }
			else if (type == 'BOOLEAN')
				return { name, convert: x => x == null ? null : x && x.toLowerCase() === 'true' ? true : false }
			else if (type == 'TIMESTAMP' || type == 'DATETIME')
				return { name, convert: x => {
					if (!x) 
						return null 
					const t = `${x}`
					if (/^[0-9]*\.{0,1}[0-9]*[eE]{0,1}[0-9]+$/.test(t))
						return /[eE]/.test(t) ? new Date(t*1000) : new Date(t)
					else
						return new Date(t)
				}}
			else if (type == 'RECORD') 
				return { name, convert: x => bigQueryResultToJson({ schema: { fields }, rows: [x]  })[0] }
			else
				return { name, convert: x => x }
		})
		
		return data.rows.filter(x => x).map(({ f }) => schema.reduce((acc, { name, convert }, idx) => {
			const v = f[idx].v
			if (Array.isArray(v))
				acc[name] = v.map(x => convert(x.v))
			else
				acc[name] = convert(v)
			return acc 
		}, {}))
	} else
		return data || []
}

module.exports = {
	cleanData,
	fitToSchema,
	schemaToFields,
	bigQueryResultToJson,
	fieldsToSchema
}